From: Pavel Emelyanov <xemul@openvz.org>
Date: Fri, 10 Apr 2009 11:22:09 +0000 (+0400)
Subject: namespaces: ban netns creation even for ve0's root
X-Git-Url: http://git.openvz.org/?p=linux-2.6.27-openvz;a=commitdiff_plain;h=5728d99a5ae464d533487ef5874a3f9ea9ff2097

namespaces: ban netns creation even for ve0's root

netnamespace do not work with sysfs yet, so creating one
can be fatal

http://bugzilla.openvz.org/show_bug.cgi?id=1234

Signed-off-by: Pavel Emelyanov <xemul@openvz.org>
---

diff --git a/kernel/nsproxy.c b/kernel/nsproxy.c
index 052b894..25288ad 100644
--- a/kernel/nsproxy.c
+++ b/kernel/nsproxy.c
@@ -142,9 +142,20 @@ int copy_namespaces(unsigned long flags, struct task_struct *tsk,
 				CLONE_NEWUSER | CLONE_NEWPID | CLONE_NEWNET)))
 		return 0;
 
-	if (!capable(CAP_SYS_ADMIN) && !force_admin) {
-		err = -EPERM;
-		goto out;
+	if (!force_admin) {
+		if (!capable(CAP_SYS_ADMIN)) {
+			err = -EPERM;
+			goto out;
+		}
+
+		/*
+		 * netns-vs-sysfs is deadly broken, thus new namespace
+		 * (even in ve0) can bring the node down
+		 */
+		if (flags & CLONE_NEWNET) {
+			err = -EINVAL;
+			goto out;
+		}
 	}
 
 	/*
@@ -221,6 +232,9 @@ int unshare_nsproxy_namespaces(unsigned long unshare_flags,
 	if (!capable(CAP_SYS_ADMIN))
 		return -EPERM;
 
+	if (unshare_flags & CLONE_NEWNET)
+		return -EINVAL;
+
 	*new_nsp = create_new_namespaces(unshare_flags, current,
 				new_fs ? new_fs : current->fs);
 	if (IS_ERR(*new_nsp)) {
